### PR TITLE
[SPARK-22435][SQL] Support processing array and map type using script

### PR DESCRIPTION
**Pull request by jinxing64**
## What changes were proposed in this pull request?

Currently, It is not supported to use script(e.g. python) to process array type or map type, it will complain with below message:
`org.apache.spark.sql.catalyst.expressions.UnsafeArrayData cannot be cast to [Ljava.lang.Object`
`org.apache.spark.sql.catalyst.expressions.UnsafeMapData cannot be cast to java.util.Map`

This pr proposes to support it by using `DelimitedJSONSerDe`
This pr also fixes a bug -- when using input row format with script, no data will be produced from `ScriptTransformationExec`.

## How was this patch tested?

Tests added.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 1]
```diff
@@ -1454,22 +1454,24 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
...
+        defaultConfigValue: String,
+        isInFormat: Boolean): Format = fmt match {
       case c: RowFormatDelimitedContext =>
         // TODO we should use the visitRowFormatDelimited function here. However HiveScriptIOSchema
         // expects a seq of pairs in which the old parsers' token names are used as keys.
         // Transforming the result of visitRowFormatDelimited would be quite a bit messier than
         // retrieving the key value pairs ourselves.
         def entry(key: String, value: Token): Seq[(String, String)] = {
-          Option(value).map(t => key -> t.getText).toSeq
+          Option(value).toSeq.map(x => key -> string(x))
```
Comment (jinxing64) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: `t.getText` doesn't work, we need to process the token. e.g. remove the quote

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 2]
```diff
@@ -1485,21 +1487,27 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
...
+          conf.getConfString("hive.script.serde",
+            "org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe")
+        }
         val props = Seq("field.delim" -> "\t")
         val recordHandler = Option(conf.getConfString(configKey, defaultConfigValue))
         (Nil, Option(name), props, recordHandler)
     }
 
-    val (inFormat, inSerdeClass, inSerdeProps, reader) =
+    val (inFormat, inSerdeClass, inSerdeProps, writer) =
```
Comment (jinxing64) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: `writer` writes records into the input stream of script. Isn't it should be initialized with input format?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 3]
```diff
@@ -1485,21 +1487,27 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
 
       case null =>
         // Use default (serde) format.
-        val name = conf.getConfString("hive.script.serde",
-          "org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe")
+        val name = if (isInFormat) {
+          conf.getConfString("hive.script.serde",
+            "org.apache.hadoop.hive.serde2.DelimitedJSONSerDe")
```
Comment (jinxing64) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: `DelimitedJSONSerDe` doesn't support deserialize.
Use `DelimitedJSONSerDe` for input SerDe and `LazySimpleSerDe` for output SerDe.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala [hunk 4]
```diff
@@ -267,6 +268,33 @@ private class ScriptTransformationWriterThread(
   /** Contains the exception thrown while writing the parent iterator to the external process. */
   def exception: Option[Throwable] = Option(_exception)
 
+  private def buildJSONString(sb: StringBuilder, obj: Any): Unit = {
```
Comment (jinxing64) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala]: TODO: build json string for more types

**Comment by gatorsmile** [comment]
Thanks for working on it. Will review it next week.

**Comment by jinxing64** [comment]
@gatorsmile 
(Very gentle ping)
Could you please give some comments when you have time :)
Thanks you so much :)

**Pull request closed without merging**